lib/ansible/modules/network/iosxr/iosxr_interface.py

```diff
@@ -125,240 +150,453 @@
+            ('line-state', {'xpath': 'interface-properties/data-nodes/data-node/system-view/interfaces/interface/line-state', 'tag': True}),
+        ])
+
+        _rate_filter = build_xml('infra-statistics', xmap=self._data_rate_meta, params=self._want, opcode="filter")
+        out = get_oper(self._module, filter=_rate_filter)
+        data_rate_list = etree_findall(out, 'interface')
+        data_rate_map = dict()
+        for item in data_rate_list:
+            data_rate_map.update({etree_find(item, 'interface-name').text: dict()})
+            data_rate_map[etree_find(item, 'interface-name').text].update({'input-data-rate': etree_find(item, 'input-data-rate').text,
+                                                                           'output-data-rate': etree_find(item, 'output-data-rate').text})
+
+        _line_state_filter = build_xml('interface-properties', xmap=self._line_state_meta, params=self._want, opcode="filter")
+        out = get_oper(self._module, filter=_line_state_filter)
+        line_state_list = etree_findall(out, 'interface')
+        line_state_map = dict()
+        for item in line_state_list:
+            line_state_map.update({etree_find(item, 'interface-name').text: etree_find(item, 'line-state').text})
+
+        for want_item in self._want:
+            want_state = want_item.get('state')
+            want_tx_rate = want_item.get('tx_rate')
+            want_rx_rate = want_item.get('rx_rate')
+            if want_state not in ('up', 'down') and not want_tx_rate and not want_rx_rate:
+                continue
+
+            if self._result['changed']:
+                sleep(want_item['delay'])
+
+            if want_state in ('up', 'down'):
+                if want_state not in line_state_map[want_item['name']]:
+                    failed_conditions.append('state ' + 'eq({!s})'.format(want_state))
+
+            if want_tx_rate:
+                if want_tx_rate != data_rate_map[want_item['name']]['output-data-rate']:
+                    failed_conditions.append('tx_rate ' + want_tx_rate)
+
+            if want_rx_rate:
+                if want_rx_rate != data_rate_map[want_item['name']]['input-data-rate']:
+                    failed_conditions.append('rx_rate ' + want_rx_rate)
+
+        if failed_conditions:
+            msg = 'One or more conditional statements have not been satisfied'
+            self._module.fail_json(msg=msg, failed_conditions=failed_conditions)
+
+    def run(self):
+        self.map_params_to_obj()
+        self.map_obj_to_xml_rpc()
+        self.check_declarative_intent_params()
+        return self._result
 
 
 def main():
     """ main entry point for module execution
     """
     element_spec = dict(
-        name=dict(),
-        description=dict(),
-        speed=dict(),
+        name=dict(type='str'),
+        description=dict(type='str'),
+        speed=dict(choices=['10', '100', '1000']),
         mtu=dict(),
```

Comment:
int?

Reply:
After re-looking the code, feels that having them as default `str` is better as the parsed response from devices will have `str`, that will help us avoid conversion from `int` --> `str` again during command or netconf xml creation as well as comparisons.